xlators/storage/posix/src/posix-helpers.c

```diff
@@ -445,32 +446,26 @@ _posix_xattr_get_set(dict_t *xattr_req, char *key, data_t *data,
 
         /* file content request */
         req_size = data_to_uint64(data);
-        if (req_size >= filler->stbuf->ia_size) {
+        if (req_size >= filler->stbuf->ia_size && filler->stbuf->ia_size > 0) {
+#ifdef GF_LINUX_HOST_OS
+            _fd = open(filler->real_path, O_RDONLY | O_NOATIME);
+#else /* O_NOATIME does not exist on BSD */
             _fd = open(filler->real_path, O_RDONLY);
+#endif
             if (_fd == -1) {
                 gf_msg(filler->this->name, GF_LOG_ERROR, errno,
                        P_MSG_XDATA_GETXATTR, "Opening file %s failed",
                        filler->real_path);
                 goto err;
             }
 
-            /*
-             * There could be a situation where the ia_size is
-             * zero. GF_CALLOC will return a pointer to the
-             * memory initialized by gf_mem_set_acct_info.
-             * This function adds a header and a footer to
-             * the allocated memory.  The returned pointer
-             * points to the memory just after the header, but
-             * when size is zero, there is no space for user
-             * data. The memory can be freed by calling GF_FREE.
-             */
-            databuf = GF_CALLOC(1, filler->stbuf->ia_size, gf_posix_mt_char);
+            databuf = GF_MALLOC(filler->stbuf->ia_size, gf_posix_mt_char);
             if (!databuf) {
                 goto err;
             }
 
-            ret = sys_read(_fd, databuf, filler->stbuf->ia_size);
-            if (ret == -1) {
+            read_len = sys_read(_fd, databuf, filler->stbuf->ia_size);
+            if (read_len <= 0) {
```

Comment:
If we consider 0 as an error and we treat it exactly as the case where it's < 0, then errno may contain a misleading value. Either we should consider 0 as a success, or handle it differently to not log incorrect information.

Reply:
OK, will change back to <0